File: src/test/java/com/dg/watcher/validation/InputValidationTest.java
package com.dg.watcher.validation;

import hudson.FilePath;
import hudson.model.AbstractProject;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import java.io.IOException;
import static com.dg.watcher.validation.InputValidation.validateCustomPathToApk;
import static com.dg.watcher.validation.InputValidation.validateThresholdInMb;
import static hudson.util.FormValidation.ok;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;


public class InputValidationTest {
    @Rule
    public TemporaryFolder tempDir = new TemporaryFolder();


    @Before
    public void setUp() throws Exception {
        createTempApkFolder();
    }

    @Test
    public void should_allow_a_positive_threshold() {
        assertThat(validateThresholdInMb("2.5"), is(equalTo(ok())));
    }

    @Test
    public void should_indicate_a_negative_threshold_with_an_error_message() {
        assertThat(validateThresholdInMb("-2.5").getMessage(),
                is(equalTo("The threshold can not be negative.")));
    }

    @Test
    public void should_indicate_a_non_integral_threshold_with_an_error_message() {
        assertThat(validateThresholdInMb("abc").getMessage(),
                is(equalTo("The threshold must be a floating point number.")));
    }

    @Test
    public void should_allow_an_empty_custom_path_to_the_apk() {
        assertThat(validateCustomPathToApk("", mockProject()), is(equalTo(ok())));
    }

    @Test
    public void should_allow_an_existing_custom_path_to_the_apk() {
        assertThat(validateCustomPathToApk("temp_apk_folder", mockProject()), is(equalTo(ok())));
    }

    @Test
    public void should_indicate_a_invalid_custom_path_to_the_apk_with_an_error_message() {
        assertThat(validateCustomPathToApk("not/existing/path", mockProject()).getMessage(),
                is(equalTo("The provided path does not exist.")));
    }

    private void createTempApkFolder() throws IOException {
        tempDir.newFolder("temp_apk_folder");
    }

    private AbstractProject mockProject() {
        AbstractProject project = mock(AbstractProject.class);

        when(project.getSomeWorkspace()).thenReturn(new FilePath(tempDir.getRoot()));

        return project;
    }
}

File: src/test/kotlin/com/dg/watcher/validation/InputValidationTest.kt
package com.dg.watcher.validation

import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import hudson.FilePath
import hudson.model.AbstractProject
import hudson.util.FormValidation.ok
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder


class InputValidationTest {
    @Rule @JvmField
    var tempDir = TemporaryFolder()


    @Before
    fun setUp() {
        createTempApkFolder()
    }

    @Test
    fun `Should allow a positive threshold`() =
            assertThat(validateThresholdInMb("2.5"), `is`(equalTo(ok())))

    @Test
    fun `Should indicate a negative threshold with an error message`() =
            assertThat(validateThresholdInMb("-2.5").message, `is`(equalTo("The threshold cannot be negative.")))

    @Test
    fun `Should indicate a non integral threshold with an error message`() =
            assertThat(validateThresholdInMb("abc").message, `is`(equalTo("The threshold must be a floating point number.")))

    @Test
    fun `Should allow an empty custom path to the apk`() =
            assertThat(validateCustomPathToApk("", mockProject()), `is`(equalTo(ok())))

    @Test
    fun `Should allow an existing custom path to the apk`() =
            assertThat(validateCustomPathToApk("temp_apk_folder", mockProject()), `is`(equalTo(ok())))

    @Test
    fun `Should indicate a invalid custom path to the apk with an error message`() =
            assertThat(validateCustomPathToApk("not/existing/path", mockProject()).message,
                    `is`(equalTo("The specified path does not exist.")))

    private fun createTempApkFolder() = tempDir.newFolder("temp_apk_folder")

    private fun mockProject() = mock<AbstractProject<*, *>> {
        on { getSomeWorkspace() } doReturn FilePath(tempDir.root)
    }
}